File: AdventOfCodeCSharp/Day10.cs
using System;
using System.Text;

namespace AdventOfCodeCSharp
{
    public class Day10
    {
        private string _input;
        public int Answer2()
        {
            _input = "3113322113";
            for (int i = 0; i < 50; i++)
            {
                _input = Iterate(_input);
                Console.WriteLine($"current loop: {i}");
            }
            return _input.Length;
        }

        private string Iterate(string input)
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < input.Length; i++)
            {
                int count = 1;
                char c = input[i];
                while (i + 1 < input.Length && c == input[i + 1])
                {
                    i++;
                    count++;
                }
                sb.Append(count);
                sb.Append(c);

            }
            return sb.ToString();
        }
    }
}

File: AdventOfCode/Day10.fs
module Day10
    
    open System

    let private insert e state = 
        match state with
        | cur::rest ->
            match cur with
            | h::_ when e = h ->  (e::cur)::rest // add e to current list
            | _ -> [e]::state   // start a new list 
        | _ -> [[e]]
        
    let private nextSeq input = 
        let numbers = 
            input 
            |> Seq.map (fun n -> Int32.Parse(n.ToString())) 
            |> Seq.toList

        let runs = List.foldBack insert numbers []

        let nextBlock currentBlock (run: list<int>) =
            let runLength = run.Length
            let num = List.head run
            currentBlock + runLength.ToString() + num.ToString()

        List.fold (fun acc n -> nextBlock acc n) "" runs
        
        
    let private initialInput = "3113322113"

    let rec private finalSeq current count = 
        //Console.WriteLine(current.ToString() + " current loop:" + count.ToString()) |> ignore   
        Console.WriteLine("current loop:" + count.ToString()) |> ignore 
        if count = 50 then current
        else 
            let next = nextSeq current
            let newCount = count+1
            finalSeq next newCount

    let private final = finalSeq initialInput 0

    let answer = final.Length          

